### PR TITLE
Dynamically detect export file type

### PR DESCRIPTION
**Pull request by AyushDharDubey**
**Problem**

Dynamically detect export file type and set the value of coerce_to_string as False if the file type has native data type support such as excel file etc. (#1844)

**Solution**

Derieve the value of coerce_to_string for given file type in admin.py get_export_data() method and pass it down upto Field.export() method using key worded arguments.

Also in Widget class, instead of setting coerce_to_string parameter's default value as True, use None as default value. Then check if coerce_to_string's value is other than None, that means an explicit value has been defined by user, in that case use the user's coerce_to_string. 

Otherwise, set coerce_to_string_is_explicitly_defined flag as False and then use derived value of coerce_to_string from coerce_to_string in Fields.export() method.

**Acceptance Criteria**

Added test

### CODE REVIEW

**Review by matthewhegarty — COMMENTED** [review]
Thanks for submitting.
I made some modifications to this and pushed in #1939.  
I introduced a new kwarg value which gets passed all the way into the widget class.
See what you think.

File: import_export/fields.py [hunk 1]
```diff
@@ -139,12 +139,14 @@ def save(self, instance, row, is_m2m=False, **kwargs):
...
                         getattr(instance, attrs[-1]).set(cleaned)
 
-    def export(self, instance):
+    def export(self, instance, **kwargs):
         """
         Returns value from the provided instance converted to export
         representation.
         """
         value = self.get_value(instance)
+        if not self.widget.coerce_to_string_is_explicitly_defined:
```
[the remark this answers is not in the document]
Comment (AyushDharDubey) [reply import_export/fields.py:148]: Yes, I agree. Your approach is a neat way of handling it.

File: import_export/formats/base_formats.py [hunk 2]
```diff
@@ -229,3 +229,15 @@ def export_data(self, dataset, **kwargs):
     )
     if fmt.is_available()
 ]
+
+#: These are the formats which support different data types (such as datetime
+#: and numbers) for which coerce_to_string is to be set false dynamically.
+DATA_TYPE_RICH_FORMATS = [
```
[the remark this answers is not in the document]
Comment (AyushDharDubey) [reply import_export/formats/base_formats.py:235]: Yes, DATA_TYPE_RICH_FORMATS sounded odd to me too, but I was also facing the same struggle in finding a good name.

'PROPRIETARY_FORMATS' is also not the best one, but I can't think of anything better
So I think for now we can go with 'PROPRIETARY_FORMATS'

[the remark this answers is not in the document]
Comment (matthewhegarty) [reply import_export/formats/base_formats.py:235]: Yes, agreed.  I think I'll go with `BINARY_FORMATS` even though it's not completely accurate, at least it is descriptive and consistent.

**Comment by matthewhegarty** [comment]
Superseded by #1939

**Pull request closed without merging**